python/ray/data/_internal/execution/operators/union_operator.py

```diff
@@ -0,0 +1,112 @@
+from typing import List, Optional
+
+from ray.data._internal.execution.interfaces import PhysicalOperator, RefBundle
+from ray.data._internal.execution.operators.base_physical_operator import NAryOperator
+from ray.data._internal.stats import StatsDict
+
+
+class UnionOperator(NAryOperator):
+    """An operator that combines blocks from input operators."""
+
+    def __init__(
+        self,
+        *input_ops: PhysicalOperator,
+    ):
+        """Create a UnionOperator.
+
+        Args:
+            input_ops: Operators generating input data for this operator to union.
+        """
+        self._input_buffers: List[List[RefBundle]] = [[] for _ in range(len(input_ops))]
+        # Whether the input dependency still has blocks remaining.
+        self._input_dependency_has_next: List[bool] = [
+            True for _ in range(len(input_ops))
+        ]
+        # The index of the input dependency that is currently the source of
+        # the output buffer. New inputs from this input dependency will be added
+        # directly to the output buffer.
+        self._input_idx_to_output = None
```

Comment:
I think what @ericl suggested is that when not preserving orders, blocks from different input deps can interwoven. That is, we can output whatever is passed into `add_input` immediately. @ericl correct me if it's not true. 

And please also add a `preserve_order` flag and support order preservation mode.

Reply:
Yeah, I think we can avoid buffering blocks completely when not preserving order. That could significantly reduce intermediate memory usage.